### ISSUE TITLE
Traceback when including a repo that doesn't need to be published in tool.conductor.repos

### ISSUE THREAD

**Issue by soininen** [spine-tools/spine-conductor#22]
I had this situation in my hands:

spinetoolbox - commits since publication
spine_items - commits since publication
spine_engine - HEAD at latest version tag
spinedb_api = ,commits since publication

I tried to use my `release.toml` without modifications but when running `conduct release --bump patch -c release.toml`, I get a Traceback when selecting files to commit for spine_engine. The gist is, expectedly, `fatal: tag '0.23.3' already exists`.

Instead of Traceback, I wish conductor would notice that spine_engine does not need to be published this time and would continue as if nothing happened. Perhaps other option would be to reset changes to the other repositories.

**Comment by suvayu** [comment]
hmm, interesting. I think the current logic assumes that because of cyclic dependency. But it could be relaxed for only the packages where there is a cyclic dependency.  Is the dependency diagram below correct?
```mermaid
classDiagram
    spinetoolbox <|-- spine_items
    spinetoolbox <|-- spine_engine
    spinetoolbox <|-- spinedb_api

    spine_items <|-- spinedb_api
    spine_items <|-- spine_engine
    spine_items <|-- spinetoolbox

    spine_engine <|-- spinedb_api
```

For now I think you should be able to workaround this by explicitly including all the packages except `spine_engine`:
```shell
conduct release --bump patch -c release.toml --only spinetoolbox --only spine_items --only spinedb_api
``` 
The [documentation][1] for this particular option isn't particularly prominent :shrug:. It's also quite verbose, I guess I could add an option `--exclude pkgname`. WDYT?

[1]: https://github.com/spine-tools/spine-conductor/blob/master/docs/making-a-release.md#how-to-limit-the-new-release-to-a-subset-of-packages

**Comment by suvayu** [comment]
I added `--exclude` in 42952e5

**Comment by suvayu** [comment]
@soininen Do you have the traceback saved somewhere? It seems this case is already considered: https://github.com/spine-tools/spine-conductor/blob/0d4ac5ecc973dd08f0ff69b3ad9c21b63f8e6385/orchestra/release.py#L282-L284

The traceback would help in understanding what is going on.

**Comment by soininen** [comment]
I completely missed the `--only` flag, sorry about that! Having `--exclude` wouldn't hurt --- in this case it is more convenient than three `--only`s.

Unfortunately, I did not save the Traceback anywhere. I will try to think a way to reproduce it.

**Comment by soininen** [comment]
I managed to recreate the Traceback. The ability to specify `tool.conductor.branches` was very helpful here as I did not have to push bogus commits and tags to `master` branches.

```
Repository: spine-engine
## master...origin/master
 [1] M pyproject.toml
Select the files to add (comma/space separated list): 1
Creating tag: 0.23.3 @ 243a756
┌─────────────────────────────── Traceback (most recent call last) ────────────────────────────────┐
│ spine-conductor\orchestra\__main__.py:56 in release                   │
│                                                                                                  │
│    53 ):                                                                                         │
│    54 │   """Tag releases for all packages."""                                                   │
│    55 │   conf = read_conf(f"{config}")                                                          │
│ >  56 │   make_release(conf, bump_version, output, only)                                         │
│    57                                                                                            │
│    58                                                                                            │
│    59 @cli.command()                                                                             │
│                                                                                                  │
│ ┌─────────────────────────────────────────── locals ───────────────────────────────────────────┐ │
│ │ bump_version = <VersionPart.patch: 'patch'>                                                  │ │
│ │         conf = {                                                                             │ │
│ │                │   'packagename_regex': 'spine(toolbox|(db){0,1}[_-][a-z]+)',                │ │
│ │                │   'pkgname_re': re.compile('spine(toolbox|(db){0,1}[_-][a-z]+)'),           │ │
│ │                │   'repos': {'spinetoolbox': '.', 'spine_items': '../spine-items',           │ │
│ │                'spine_engine': '../spine-engine', 'spinedb_api': '../Spine-Database-API'},   │ │
│ │                │   'dependency_graph': {'spinetoolbox': ['spine_items', 'spine_engine',      │ │
│ │                'spinedb_api'], 'spine_items': ['spinetoolbox', 'spine_engine',               │ │
│ │                'spinedb_api'], 'spine_engine': ['spinedb_api'], 'spinedb_api': []},          │ │
│ │                │   'default_branch': 'master',                                               │ │
│ │                │   'branches': {'spinetoolbox': 'reproduce_conductor_traceback',             │ │
│ │                'spine_items': 'reproduce_conductor_traceback', 'spine_engine': 'master',     │ │
│ │                'spinedb_api': 'reproduce_conductor_traceback'},                              │ │
│ │                │   'workflow': {'repo': 'spine-tools/spine-conductor', 'file':               │ │
│ │                'test-n-publish.yml'}                                                         │ │
│ │                }                                                                             │ │
│ │       config = WindowsPath('release.toml')                                                   │ │
│ │          ctx = <click.core.Context object at 0x000001A6EA5D7620>                             │ │
│ │         only = []                                                                            │ │
│ │       output = WindowsPath('pkgtags.json')                                                   │ │
│ └──────────────────────────────────────────────────────────────────────────────────────────────┘ │
│                                                                                                  │
│ spine-conductor\orchestra\release.py:324 in make_release              │
│                                                                                                  │
│   321 │   │   console.print(format_exc(err, "Aborting!"))                                        │
│   322 │   │   ErrorCodes.BRANCH_ERR.exit()                                                       │
│   323 │                                                                                          │
│ > 324 │   summary = create_tags(config["repos"], bump_version)                                   │
│   325 │   json_str = json.dumps(summary, indent=4)                                               │
│   326 │   output.write_text(json_str)                                                            │
│   327 │   msg = "\n[underline]Package Tags summary[/underline] "                                 │
│                                                                                                  │
│ ┌─────────────────────────────────────────── locals ───────────────────────────────────────────┐ │
│ │     branches = {'spinetoolbox': 'reproduce_conductor_traceback', 'spine_items':              │ │
│ │                'reproduce_conductor_traceback', 'spine_engine': 'master', 'spinedb_api':     │ │
│ │                'reproduce_conductor_traceback'}                                              │ │
│ │ bump_version = <VersionPart.patch: 'patch'>                                                  │ │
│ │       config = {                                                                             │ │
│ │                │   'packagename_regex': 'spine(toolbox|(db){0,1}[_-][a-z]+)',                │ │
│ │                │   'pkgname_re': re.compile('spine(toolbox|(db){0,1}[_-][a-z]+)'),           │ │
│ │                │   'repos': {'spinetoolbox': '.', 'spine_items': '../spine-items',           │ │
│ │                'spine_engine': '../spine-engine', 'spinedb_api': '../Spine-Database-API'},   │ │
│ │                │   'dependency_graph': {'spinetoolbox': ['spine_items', 'spine_engine',      │ │
│ │                'spinedb_api'], 'spine_items': ['spinetoolbox', 'spine_engine',               │ │
│ │                'spinedb_api'], 'spine_engine': ['spinedb_api'], 'spinedb_api': []},          │ │
│ │                │   'default_branch': 'master',                                               │ │
│ │                │   'branches': {'spinetoolbox': 'reproduce_conductor_traceback',             │ │
│ │                'spine_items': 'reproduce_conductor_traceback', 'spine_engine': 'master',     │ │
│ │                'spinedb_api': 'reproduce_conductor_traceback'},                              │ │
│ │                │   'workflow': {'repo': 'spine-tools/spine-conductor', 'file':               │ │
│ │                'test-n-publish.yml'}                                                         │ │
│ │                }                                                                             │ │
│ │         only = []                                                                            │ │
│ │       output = WindowsPath('pkgtags.json')                                                   │ │
│ └──────────────────────────────────────────────────────────────────────────────────────────────┘ │
│                                                                                                  │
│ spine-conductor\orchestra\release.py:305 in create_tags               │
│                                                                                                  │
│   302 │   │   console.print(                                                                     │
│   303 │   │   │   f"Creating tag: {next_version} @ [yellow]{repo.head.commit.hexsha[:7]}"        │
│   304 │   │   )                                                                                  │
│ > 305 │   │   repo.create_tag(next_version)                                                      │
│   306 │   │   summary[remote_name(repo)] = next_version                                          │
│   307 │   return summary                                                                         │
│   308                                                                                            │
│                                                                                                  │
│ ┌─────────────────────────────────────────── locals ───────────────────────────────────────────┐ │
│ │ _next_versions = ['0.7.4', '0.21.5', '0.23.3', '0.30.5']                                     │ │
│ │         _repos = [                                                                           │ │
│ │                  │   <git.repo.base.Repo                                                     │ │
│ │                  'Spine-Toolbox\\.git'>,                      │ │
│ │                  │   <git.repo.base.Repo                                                     │ │
│ │                  'spine-items\\.git'>,                        │ │
│ │                  │   <git.repo.base.Repo                                                     │ │
│ │                  spine-engine\\.git'>,                       │ │
│ │                  │   <git.repo.base.Repo                                                     │ │
│ │                  'Spine-Database-API\\.git'>                  │ │
│ │                  ]                                                                           │ │
│ │          _tags = ['v0.4.2-alpha.0', '0.9.0', '0.6.0-beta.0', 'v0.0.9']                       │ │
│ │          added = 1                                                                           │ │
│ │   bump_version = <VersionPart.patch: 'patch'>                                                │ │
│ │       modified = []                                                                          │ │
│ │            msg = 'Release 0.23.3\n\n\n'                                                      │ │
│ │   next_version = '0.23.3'                                                                    │ │
│ │  next_versions = {                                                                           │ │
│ │                  │   'spinetoolbox': '0.7.4',                                                │ │
│ │                  │   'spine_items': '0.21.5',                                                │ │
│ │                  │   'spine_engine': '0.23.3',                                               │ │
│ │                  │   'spinedb_api': '0.30.5'                                                 │ │
│ │                  }                                                                           │ │
│ │            pkg = 'spine_engine'                                                              │ │
│ │           pkgs = {                                                                           │ │
│ │                  │   'spinetoolbox': (                                                       │ │
│ │                  │   │   <git.repo.base.Repo                                                 │ │
│ │                  'Spine-Toolbox\\.git'>,                      │ │
│ │                  │   │   'v0.4.2-alpha.0',                                                   │ │
│ │                  │   │   '0.7.4'                                                             │ │
│ │                  │   ),                                                                      │ │
│ │                  │   'spine_items': (                                                        │ │
│ │                  │   │   <git.repo.base.Repo                                                 │ │
│ │                  'spine-items\\.git'>,                        │ │
│ │                  │   │   '0.9.0',                                                            │ │
│ │                  │   │   '0.21.5'                                                            │ │
│ │                  │   ),                                                                      │ │
│ │                  │   'spine_engine': (                                                       │ │
│ │                  │   │   <git.repo.base.Repo                                                 │ │
│ │                  'spine-engine\\.git'>,                       │ │
│ │                  │   │   '0.6.0-beta.0',                                                     │ │
│ │                  │   │   '0.23.3'                                                            │ │
│ │                  │   ),                                                                      │ │
│ │                  │   'spinedb_api': (                                                        │ │
│ │                  │   │   <git.repo.base.Repo                                                 │ │
│ │                  'Spine-Database-API\\.git'>,                 │ │
│ │                  │   │   'v0.0.9',                                                           │ │
│ │                  │   │   '0.30.5'                                                            │ │
│ │                  │   )                                                                       │ │
│ │                  }                                                                           │ │
│ │           repo = <git.repo.base.Repo 'spine-engine\\.git'>    │ │
│ │     repo_paths = {'spinetoolbox': '.', 'spine_items': '../spine-items', 'spine_engine':      │ │
│ │                  '../spine-engine', 'spinedb_api': '../Spine-Database-API'}                  │ │
│ │        summary = {'Spine-Toolbox': '0.7.4', 'spine-items': '0.21.5'}                         │ │
│ │            tag = '0.6.0-beta.0'                                                              │ │
│ └──────────────────────────────────────────────────────────────────────────────────────────────┘ │
│                                                                                                  │
│ Lib\site-packages\git\repo\base.py:551 in │
│ create_tag                                                                                       │
│                                                                                                  │
│    548 │   │                                                                                     │
│    549 │   │   :return: :class:`~git.refs.tag.TagReference` object                               │
│    550 │   │   """                                                                               │
│ >  551 │   │   return TagReference.create(self, path, ref, message, force, **kwargs)             │
│    552 │                                                                                         │
│    553 │   def delete_tag(self, *tags: TagReference) -> None:                                    │
│    554 │   │   """Delete the given tag references"""                                             │
│                                                                                                  │
│ ┌────────────────────────────────────── locals ──────────────────────────────────────┐           │
│ │   force = False                                                                    │           │
│ │  kwargs = {}                                                                       │           │
│ │ message = None                                                                     │           │
│ │    path = '0.23.3'                                                                 │           │
│ │     ref = 'HEAD'                                                                   │           │
│ │    self = <git.repo.base.Repo 'spine-engine\\.git'> │           │
│ └────────────────────────────────────────────────────────────────────────────────────┘           │
│                                                                                                  │
│ Lib\site-packages\git\refs\tag.py:131 in  │
│ create                                                                                           │
│                                                                                                  │
│   128 │   │                                                                                      │
│   129 │   │   args = (path, reference)                                                           │
│   130 │   │                                                                                      │
│ > 131 │   │   repo.git.tag(*args, **kwargs)                                                      │
│   132 │   │   return TagReference(repo, "%s/%s" % (cls._common_path_default, path))              │
│   133 │                                                                                          │
│   134 │   @classmethod                                                                           │
│                                                                                                  │
│ ┌─────────────────────────────────────── locals ───────────────────────────────────────┐         │
│ │      args = ('0.23.3', 'HEAD')                                                       │         │
│ │       cls = <class 'git.refs.tag.TagReference'>                                      │         │
│ │     force = False                                                                    │         │
│ │    kwargs = {}                                                                       │         │
│ │    logmsg = None                                                                     │         │
│ │      path = '0.23.3'                                                                 │         │
│ │ reference = 'HEAD'                                                                   │         │
│ │      repo = <git.repo.base.Repo 'spine-engine\\.git'> │         │
│ └──────────────────────────────────────────────────────────────────────────────────────┘         │
│                                                                                                  │
│ Lib\site-packages\git\cmd.py:800 in       │
│ <lambda>                                                                                         │
│                                                                                                  │
│    797 │   │   """                                                                               │
│    798 │   │   if name[0] == "_":                                                                │
│    799 │   │   │   return LazyMixin.__getattr__(self, name)                                      │
│ >  800 │   │   return lambda *args, **kwargs: self._call_process(name, *args, **kwargs)          │
│    801 │                                                                                         │
│    802 │   def set_persistent_git_options(self, **kwargs: Any) -> None:                          │
│    803 │   │   """Specify command line options to the git executable for subsequent              │
│                                                                                                  │
│ ┌────────────────────── locals ───────────────────────┐                                          │
│ │   args = ('0.23.3', 'HEAD')                         │                                          │
│ │ kwargs = {}                                         │                                          │
│ │   name = 'tag'                                      │                                          │
│ │   self = <git.cmd.Git object at 0x000001A6EA5E6140> │                                          │
│ └─────────────────────────────────────────────────────┘                                          │
│                                                                                                  │
│ Lib\site-packages\git\cmd.py:1386 in      │
│ _call_process                                                                                    │
│                                                                                                  │
│   1383 │   │   call.append(dashify(method))                                                      │
│   1384 │   │   call.extend(args_list)                                                            │
│   1385 │   │                                                                                     │
│ > 1386 │   │   return self.execute(call, **exec_kwargs)                                          │
│   1387 │                                                                                         │
│   1388 │   def _parse_object_header(self, header_line: str) -> Tuple[str, str, int]:             │
│   1389 │   │   """                                                                               │
│                                                                                                  │
│ ┌────────────────────────────── locals ──────────────────────────────┐                           │
│ │                  args = ('0.23.3', 'HEAD')                         │                           │
│ │             args_list = ['0.23.3', 'HEAD']                         │                           │
│ │                  call = ['git', 'tag', '0.23.3', 'HEAD']           │                           │
│ │           exec_kwargs = {}                                         │                           │
│ │              ext_args = ['0.23.3', 'HEAD']                         │                           │
│ │ insert_after_this_arg = None                                       │                           │
│ │                kwargs = {}                                         │                           │
│ │                method = 'tag'                                      │                           │
│ │              opt_args = []                                         │                           │
│ │           opts_kwargs = {}                                         │                           │
│ │                  self = <git.cmd.Git object at 0x000001A6EA5E6140> │                           │
│ └────────────────────────────────────────────────────────────────────┘                           │
│                                                                                                  │
│ Lib\site-packages\git\cmd.py:1183 in      │
│ execute                                                                                          │
│                                                                                                  │
│   1180 │   │   # END handle debug printing                                                       │
│   1181 │   │                                                                                     │
│   1182 │   │   if with_exceptions and status != 0:                                               │
│ > 1183 │   │   │   raise GitCommandError(redacted_command, status, stderr_value, stdout_value)   │
│   1184 │   │                                                                                     │
│   1185 │   │   if isinstance(stdout_value, bytes) and stdout_as_string:  # Could also be output  │
│   1186 │   │   │   stdout_value = safe_decode(stdout_value)                                      │
│                                                                                                  │
│ ┌─────────────────────────────────────────── locals ───────────────────────────────────────────┐ │
│ │              as_process = False                                                              │ │
│ │ cmd_not_found_exception = <class 'OSError'>                                                  │ │
│ │                 command = ['git', 'tag', '0.23.3', 'HEAD']                                   │ │
│ │                     cwd = 'spine-engine'                      │ │
│ │                     env = {                                                                  │ │
│ │                           │  <omitted>                         │ │
│ │                           }                                                                  │ │
│ │              inline_env = None                                                               │ │
│ │                 istream = None                                                               │ │
│ │      kill_after_timeout = None                                                               │ │
│ │            kill_process = <function Git.execute.<locals>.kill_process at 0x000001A6EA6177E0> │ │
│ │          max_chunk_size = 8192                                                               │ │
│ │                 newline = b'\n'                                                              │ │
│ │           output_stream = None                                                               │ │
│ │                    proc = <Popen: returncode: 128 args: ['git', 'tag', '0.23.3', 'HEAD']>    │ │
│ │        redacted_command = ['git', 'tag', '0.23.3', 'HEAD']                                   │ │
│ │                    self = <git.cmd.Git object at 0x000001A6EA5E6140>                         │ │
│ │                   shell = False                                                              │ │
│ │                  status = 128                                                                │ │
│ │            stderr_value = b"fatal: tag '0.23.3' already exists"                              │ │
│ │        stdout_as_string = True                                                               │ │
│ │             stdout_sink = -1                                                                 │ │
│ │            stdout_value = b''                                                                │ │
│ │ strip_newline_in_stdout = True                                                               │ │
│ │       subprocess_kwargs = {}                                                                 │ │
│ │      universal_newlines = False                                                              │ │
│ │         with_exceptions = True                                                               │ │
│ │    with_extended_output = False                                                              │ │
│ │             with_stdout = True                                                               │ │
│ └──────────────────────────────────────────────────────────────────────────────────────────────┘ │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘
GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git tag 0.23.3 HEAD
  stderr: 'fatal: tag '0.23.3' already exists'
```

**Comment by soininen** [comment]
>  Is the dependency diagram below correct?

Yes.

**Comment by suvayu** [comment]
Damn, I see the reason for the bug.  Basically the code checks for Spine dependencies from the pyproject.toml, which by-passes the only/exclude modifications, so it ends up failing on the excluded package.  This should be easy to fix. I'll do it tonight.

https://github.com/spine-tools/spine-conductor/blob/0d4ac5ecc973dd08f0ff69b3ad9c21b63f8e6385/orchestra/release.py#L161-L165

**Comment by suvayu** [comment]
Fixed in ea98743 & 7a71288
